lib/buffer.ex

```diff
@@ -39,3 +52,59 @@ defmodule Buffer.Queue do
 
   def size(queue), do: :queue.len(queue)
 end
+
+defmodule Buffer.InstrumentedQueue do
+  @moduledoc "An instumented FIFO queue"
+  @behaviour Buffer
+
+  defmodule Item do
+    @moduledoc false
+    defstruct [:inserted_at, :raw_item]
+  end
+
+  defmodule State do
+    @moduledoc false
+    defstruct [:impl, :buffer]
+  end
+
+  @doc "Creates a new empty queue. Can pass a module that implements the underlying `Buffer` behaviour."
+  def new(impl \\ Buffer.Queue), do: %State{impl: impl, buffer: impl.new()}
+
+  def push(%State{impl: impl, buffer: buffer} = state, raw_items) do
+    items =
+      raw_items
+      |> Enum.map(fn raw_item ->
+        %Item{inserted_at: :erlang.monotonic_time(), raw_item: raw_item}
+      end)
+
+    %State{state | buffer: impl.push(buffer, items)}
+  end
+
+  def pop(%State{impl: impl, buffer: buffer} = state, count) do
+    {items, buffer} = impl.pop(buffer, count)
+
+    raw_items =
+      items
+      |> Enum.map(fn %Item{raw_item: raw_item} -> raw_item end)
+
+    {raw_items, %State{state | buffer: buffer}}
+  end
+
+  @doc "Returns the microseconds that the item at the head of the queue has been queued for"
+  @spec queueing_us(Buffer.t()) :: integer()
+  def queueing_us(%State{impl: impl, buffer: buffer}) do
```

Comment:
Isn't this naming a bit obscure? Just with the name it's hard to know what this does.